LeanMapperQuery/IQuery.php

```diff
@@ -10,23 +10,20 @@
 
 use LeanMapper\Fluent;
 use LeanMapper\IMapper;
-use LeanMapper\Relationship;
+
 
 /**
  * @author Michal Bohuslávek
  */
 interface IQuery
 {
-	const RELATIONSHIP_TABLE = 0;
-	const TARGET_TABLE = 1;
-
 	/**
 	 * @param  Fluent  $fluent
 	 * @param  IMapper $mapper
-	 * @param  int|NULL $type
+	 * @param  IQueryTarget|NULL $target
```

Comment:
Zdá se mi, že to rozhraní `IQueryTarget` je úplně zbytečné a že existuje pouze kvůli tomu, abychom sem mohli něco napsat. Rozhodně bych ho smazal. Je to v podstatě pouze záležitost dokumentace:

```suggestion
	 * @param  mixed $target  one of QueryTarget\* classes
```

Reply:
Dalo by se říct, že zbytečné je, ale přišlo mi vhodné ho přidat proto

* aby nešlo do toho parametru nacpat úplně cokoli (číslo, string, nějaký nesouvisející objekt), aby to prostě akceptovalo jen objekt, který byl pro ten účel cíleně vytvořen (i kdyby to rozhraní mělo být úplně prázdné)
* minoritně, aby se zjednodušilo nastavení proměnné `$targetTable` na tomto řádku https://github.com/mibk/LeanMapperQuery/pull/4/files#diff-b377057ded2f35af677b38a98d2a423eR459 - s tím souvisí i ta metoda `getTargetTable` v rozhraní (i když lepší název by asi byl jen `getTable`)
* a s tou metodou `getTargetTable` a proměnnou `$targetTable` souvisí i určitý ohled na nějaké budoucí rozšiřování - aktuální řešení s rozhraním umožňuje (ať už nám nebo uživateli) si pro další specifické případy vytvořit vlastní implementaci toho rozhraní, ovlivnit `$targetTable` a pak klidně v potomkovi doladit i `$fluent`. To by bez rozhraní úplně nešlo - minimálně by nebylo možné univerzálně ovlivnit, jaké `$targetTable` se má předat do metody `apply()`

V podstatě to naplňuje to, co jsi psal v jednom z předchozích komentářů:

> Bylo by to tak i připravené na případné další rozšiřování: prostě přidáme další třídu.

Rozhraní to rozšiřování zobecňuje do té míry, aby přidání další třídy vyžadovalo co nejmíň úprav.

Když by to byl problém, můžu to rozhraní dát pryč, jen mi to přijde trochu proti srsti :)

----

Jinak s aktuální implementací jsem spokojen, myslím, že se nám podařilo dojít k optimálnímu řešení, jen ještě dořešit těch pár drobností a můžeme to vyzkoušet v praxi :)

Reply:
> * aby nešlo do toho parametru nacpat úplně cokoli (číslo, string, nějaký nesouvisející objekt), aby to prostě akceptovalo jen objekt, který byl pro ten účel cíleně vytvořen (i kdyby to rozhraní mělo být úplně prázdné)

Chápu, ale skoro se mi nechce věřit, že by tuto metodu volal někdo sám. Chybu to vyhodí tak jako tak, takže metodu nelze nějak špatně použít, že by to při vývoji fungovalo a na produkci dělalo neplechu. Lišit se to bude v podstatě jen v tom, jakou to hodí chybu.

> * minoritně, aby se zjednodušilo nastavení proměnné $targetTable na tomto řádku

Rozumím, ale myslím, že kvůli ušetření pár řádků to za to nestojí.

> * a s tou metodou `getTargetTable` a proměnnou `$targetTable` souvisí i určitý ohled na nějaké budoucí rozšiřování - aktuální řešení s rozhraním umožňuje (ať už nám nebo uživateli) si pro další specifické případy vytvořit vlastní implementaci toho rozhraní, ovlivnit `$targetTable` a pak klidně v potomkovi doladit i `$fluent`.

Vlastní implementaci toho rozhraní stejně neumožňujeme, protože limitujeme ten výčet potomků pomocí `instanceof`. Musel by tedy jedno z nich podědit, což je v podstatě hodně velké hackování. Musel by se koukat do střev kódu a něco nahackovat takovým způsobem, aby to zrovna fungovalo, jak potřebuje. V tu chvíli je snad jednodušší, aby si napsal vlastní SQL dotaz sám. *Co se týče případného doladění `$fluent`, to si může doladit hned po zavolání `applyQuery` sám.*

> To by bez rozhraní úplně nešlo - minimálně by nebylo možné univerzálně ovlivnit, jaké `$targetTable` se má předat do metody `apply()`

Tohle ani nechceme. Nechceme dát uživateli víc možností, jak to ohackovat, než je nezbytně nutné. Úplně nejlepší by bylo, kdyby součástí veřejného rozhraní `LMQuery` bylo pouze `Entity`. Třída `Query` je prostě příliš low-level na to, aby si s ní někdo hrál.

Shrnutí
--------

Nemyslím si, že je rozhraní úplně potřeba, ale pokud se Ti zdá, že tam z nějakého důvodu patří, navrhoval bych, aby (1) nemělo žádné metody a (2) bylo přesunuto do namespace `QueryTarget` pod názvem `ITarget`.

Reply:
PR jsem upravil, rozhraní ponechal prázdné a přesunul pod namespace `QueryTarget`.